src/Microsoft.OData.Core/ODataMetadataOutputContext.cs

```diff
@@ -182,6 +182,49 @@ internal override void WriteMetadataDocument()
         /// </summary>
         /// <param name="disposing">If 'true' this method is called from user code; if 'false' it is called by the runtime.</param>
         protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                try
+                {
+                    if (this.xmlWriter != null)
+                    {
+                        // In the async case the underlying stream is the async buffered stream, so we have to flush that explicitly.
+                        if (this.asynchronousOutputStream != null)
+                        {
+                            this.xmlWriter.FlushAsync().Wait();
+                            this.asynchronousOutputStream.FlushAsync().Wait();
+                            this.asynchronousOutputStream.Dispose();

```

Comment:
why do we call async version of flush, but sync version of dispose?

Reply:
Good point. This method was implemented way before we had implemented `DisposeAsync()`. But we should actually call `Flush` instead of `FlushAsync()` in this method since `FlushAsync().Wait()` is arguably more costly than just calling the synchronous `Flush()`: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#avoid-using-taskresult-and-taskwait

Reply:
:thumbsup: